### PR TITLE
DataBuffer, Averager, Normalizer, DIAYN, RND, ICM, NestCombiner, and InputPreprocessor

### PR DESCRIPTION
**Pull request by hnyu**
#416, #410, #400, #401, #402, 

Didn't know that @witwolf  was also working on supporting nested inputs #412 ... Maybe can merge the two PRs.

Also make all networks derive from the base class `Network`. 

For now the algorithms don't come with trained results. I will add after Agent is implemented.

### CODE REVIEW

File: alf/algorithms/diayn_algorithm.py [hunk 1]
```diff
@@ -108,37 +119,40 @@ def train_step(self,
         observation, skill = observations_aug
         prev_skill = state
 
+        # normalize observation for easier prediction
+        if self._observation_normalizer is not None:
+            observation = self._observation_normalizer.normalize(observation)
+
         if self._encoding_net is not None:
-            feature, _ = self._encoding_net(observation)
+            feature = self._encoding_net(observation)
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: Let's make the interface of all kinds of Network same: always have state_spec property, and forward() always have state argument, and return (output, state) tuple. The reason is to make switching between RNN and non-RNN network easier.

File: alf/algorithms/diayn_algorithm.py [hunk 2]
```diff
@@ -108,37 +119,40 @@ def train_step(self,
...
+        return AlgStep(
+            output=(),
             state=skill,
             info=DIAYNInfo(
                 reward=intrinsic_reward,
                 loss=LossInfo(
-                    loss=skill_discriminate_loss,
-                    extra=dict(
-                        skill_discriminate_loss=skill_discriminate_loss))))
+                    loss=loss, extra=dict(skill_discriminate_loss=loss))))
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: can be optimized a little bit:
```python
return AlgStep(
    output=(),
    state=skill,
    info=DIAYNInfo(
        reward=intrinsic_reward,
        loss=loss))
```
And multiplying with `valid_masks` and make the LossInfo as `calc_loss()`.
This can reduce the number of operations because of batching. And it also reduces memory of the `extra` field.

File: alf/networks/utils.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            raise ValueError('Batch dims must be non-negative.')
+        self._batch_dims = batch_dims
+        self._original_tensor_shape = None
+
+    def flatten(self, tensor):
+        """Flattens and caches the tensor's batch_dims."""
+        if self._batch_dims == 1:
+            return tensor
+        self._original_tensor_shape = tensor.shape
+        return tensor.view(-1, *tensor.shape[self._batch_dims:])
```
Comment (emailweixu) [inline alf/networks/utils.py]: It may not always possible to use `view()`.  `reshape()` is safer.  Whenever possible reshape will use view, otherwise it will copy. see https://pytorch.org/docs/stable/tensors.html#torch.Tensor.reshape

File: alf/utils/averager.py [hunk 4]
```diff
@@ -14,33 +14,36 @@
...
+        spec (TensorSpec):
 
     Returns:
         the average tensor across outer dims
     """
     outer_dims = get_outer_rank(tensor, spec)
     batch_squash = BatchSquash(outer_dims)
     tensor = batch_squash.flatten(tensor)
-    return tf.reduce_mean(tensor, axis=0)
+    return torch.mean(tensor, dim=0)
```
Comment (emailweixu) [inline alf/utils/averager.py]: tensor.mean(dim=list(range(outer_dims)))

File: alf/utils/data_buffer.py [hunk 5]
```diff
@@ -81,21 +72,18 @@ def add_batch(self, batch):
...
+        batch_size = alf.nest.get_nest_batch_size(batch)
+        with alf.device(self._device):
+            n = torch.min(self._capacity, torch.as_tensor(batch_size))
+            indices = torch.arange(self._current_pos,
+                                   self._current_pos + n) % self._capacity
+            alf.nest.map_structure(
+                lambda buf, bat: buf.__setitem__(indices, bat[-n:].detach()),
+                self._buffer, batch)
+
+            self._current_pos = (self._current_pos + n) % self._capacity
```
Comment (emailweixu) [inline alf/utils/data_buffer.py]: use `self._current_pos.copy_()` so that the object `self._current_pos` refers to does not change,  which can reduce the overhead of book keeping.

File: alf/layers.py [hunk 6]
```diff
@@ -29,6 +31,40 @@ def identity(x):
     return x
 
 
+def square(x):
```
Comment (emailweixu) [inline alf/layers.py]: seems to belong to math_ops.py

File: alf/algorithms/diayn_algorithm.py [hunk 7]
```diff
@@ -11,18 +11,19 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from collections import namedtuple
 
-import gin.tf
-import tensorflow as tf
+import gin
+import functools
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: not used

File: alf/algorithms/diayn_algorithm.py [hunk 8]
```diff
@@ -11,18 +11,19 @@
...
 
-from tf_agents.networks.network import Network
-import tf_agents.specs.tensor_spec as tensor_spec
+import torch
 
-from alf.algorithms.algorithm import Algorithm, AlgorithmStep, LossInfo
+import alf
+from alf.tensor_specs import TensorSpec, BoundedTensorSpec
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: sort imports alphabetically

File: alf/algorithms/diayn_algorithm.py (hunk 2 above)
Comment (hnyu) [reply alf/algorithms/diayn_algorithm.py]: I cannot get StepType in calc_loss() for computing valid_masks?

File: alf/layers.py [hunk 9]
```diff
@@ -29,6 +31,40 @@ def identity(x):
...
+    """torch doesn't have square."""
+    return torch.pow(x, 2)
+
+
+def batch_normalization(x, mean, variance, variance_epsilon):
```
[the remark this answers is not in the document]
Comment (hnyu) [reply alf/layers.py]: It's indeed literally "batch normalization"

File: alf/algorithms/diayn_algorithm.py [hunk 10]
```diff
@@ -37,66 +38,76 @@ class DIAYNAlgorithm(Algorithm):
...
-                 feature_spec,
-                 hidden_size=256,
+                 skill_spec,
+                 encoding_net: EncodingNetwork,
                  reward_adapt_speed=8.0,
-                 encoding_net: Network = None,
-                 discriminator_net: Network = None,
+                 observation_spec=None,
+                 hidden_size=(),
+                 hidden_activation=torch.relu,
```
[the remark this answers is not in the document]
Comment (hnyu) [reply alf/algorithms/diayn_algorithm.py]: The encoding network can handle images. To keep the API concise, I think there is no need for discriminator_net

File: alf/layers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Haichao-Zhang) [reply alf/layers.py]: I understand this function implements a specific way to normalize a batch of data, but the name ```batch_normalization``` is typically refers to the specific [operation](https://en.wikipedia.org/wiki/Batch_normalization), especially under the context of DNN. Since there are several difference with the standard version, it's better to use a different name, maybe ```batch_ standardization```.

[the remark this answers is not in the document]
Comment (hnyu) [reply alf/layers.py]: > I understand this function implements a specific way to normalize a batch of data, but the name `batch_normalization` typically refers to the specific [operation](https://en.wikipedia.org/wiki/Batch_normalization), especially under the context of DNN. Since there are several difference with the standard version, it's better to use a different name, maybe `batch_ standardization`.

That's what tf2 calls. I can change it to `normalize_along_batch_dims`.

File: alf/networks/network.py [hunk 11]
```diff
@@ -133,12 +226,49 @@ def copy(self, **kwargs):
...
         return super(Network, self).__call__(inputs, *args, **kwargs)
 
+    @property
+    def nested_input_tensor_spec(self):
+        """Return the input tensor spec BEFORE preprocessings have been applied.
+        """
+        return self._nested_input_tensor_spec
+
+    @property
+    def input_tensor_spec(self):
```
Comment (emailweixu) [inline alf/networks/network.py]: I suggest this to be the original input tensor spec. i.e. (self._nested_input_tensor_spec) to reduce confusion. And it seems there is no need to expose the tensor spec after preprocessing.

File: alf/networks/network.py [hunk 12]
```diff
@@ -133,12 +226,49 @@ def copy(self, **kwargs):
...
+                 skip_input_preprocessing=False,
+                 name="DistributionNetwork"):
+        super(DistributionNetwork, self).__init__(
+            input_tensor_spec, input_preprocessors, preprocessing_combiner,
+            skip_input_preprocessing, name)
+
+    @property
+    def output_spec(self):
+        if self._output_spec is None:
+            self._output_spec = DistributionSpec.from_distribution(
```
Comment (emailweixu) [inline alf/networks/network.py]: use `dist_utils.extract_spec()` instead in case that the output is nested.

File: alf/networks/preprocessors.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        return ret
+
+
+@gin.configurable
+class EmbeddingPreprocessor(InputPreprocessor):
+    """A preprocessor that converts the input to an embedding vector. This can
+    be used when the input is a discrete scalar, or a continuous vector to be
+    projected to a different dimension (to have the same length with other
+    vectors). Different from an `EncodingNetwork`, the input can be in the
+    original format from the environment.
```
Comment (emailweixu) [inline alf/networks/preprocessors.py]: Need to comment that this is not suitable for word embedding or discrete input with large number of discrete values. Or use [nn.Embedding](https://pytorch.org/docs/stable/nn.html#torch.nn.Embedding) to embed discrete input

File: alf/networks/network.py [hunk 14]
```diff
@@ -84,35 +91,121 @@ def _capture_init(self, *args, **kwargs):
...
+                for different inputs by configuring a gin file without changing
+                the code. For example, embedding a discrete input before concatenating
+                it to another continuous vector.
+            preprocessing_combiner (NestCombiner): preprocessing called on
+                complex inputs. Note that this combiner must also accept
+                `input_tensor_spec` as the input to compute the processed
+                tensor spec. For example, see `alf.nest.utils.NestConcat`. This
+                arg is helpful if you want to combine inputs by configuring a
+                gin file without changing the code.
+            skip_input_preprocessing (bool): If True, skip preprocessing and
```
Comment (emailweixu) [inline alf/networks/network.py]: Can't this be inferred from `input_preprocessors`? if input_preprocessors is None, then skip.

File: alf/utils/tensor_utils.py [hunk 15]
```diff
@@ -57,7 +57,7 @@ def explained_variance(ypred, y):
     Returns:
         1 - Var[y-ypred] / Var[y]
     """
-    ypred = ypred.view(-1)
-    y = y.view(-1)
+    ypred = torch.reshape(ypred, (-1, ))
+    y = torch.reshape(y, (-1, ))
```
Comment (emailweixu) [inline alf/utils/tensor_utils.py]: line 60-61 can be removed if the following two lines are changed:
```python
    vary = torch.var(y, unbiased=False)
    return 1 - torch.var(y - ypred, unbiased=False) / (vary + 1e-30)
```

File: alf/data_structures.py [hunk 16]
```diff
@@ -118,7 +118,7 @@ def _create_timestep(observation, prev_action, reward, discount, env_id,
     create_tensor = lambda t: torch.as_tensor(t).detach()
     make_tensors = lambda struct: map_structure(create_tensor, struct)
     return TimeStep(
-        step_type=step_type.view(discount.shape),
+        step_type=torch.reshape(step_type, discount.shape),
```
[the remark this answers is not in the document]
Comment (hnyu) [reply alf/data_structures.py]: I believe @le-horizon had some reason for the original `view`. This function is only used in `metrics_test`

File: alf/nest/utils.py [hunk 17]
```diff
@@ -13,54 +13,110 @@
...
+        Returns:
+            tensor (Tensor): the combined result
+        """
+        pass
+
+    def __call__(self, nested):
+        """Combine all elements according to the method defined in `combine_flat`.
+
+        Args:
+            nested (nest): a nested structure; each element can be either a
```
[the remark this answers is not in the document]
Comment (hnyu) [reply alf/nest/utils.py]: In the base class `Network __init__`, the function will automatically calculate the tensor spec after combining the nested input. So later the downstream networks can use `self._input_tensor_spec` it computes.

File: alf/data_structures.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (le-horizon) [reply alf/data_structures.py]: I think it was because step_type is a scalar, and cannot be concatenated with other tensors in the TimeStep nest.  View is necessary to avoid complaints.

File: alf/networks/encoding_networks.py [hunk 18]
```diff
@@ -79,8 +85,9 @@ def __init__(self,
                     padding=padding))
             input_channels = filters
 
-    def output_shape(self):
-        """Return the output shape given the input image size.
+    @property
+    def output_spec(self):
```
[the remark this answers is not in the document]
Comment (hnyu) [reply alf/networks/encoding_networks.py]: I think for some networks like `CriticNetwork` and `ValueNetwork`, their output spec is quite obvious. So there is no need to involve the `_test_forward`

File: alf/networks/network.py (hunk 11 above)
Comment (hnyu) [reply alf/networks/network.py]: But then `self.input_tensor_spec` and `self._input_tensor_spec` return two different things. And the latter is always needed for constructing downstream networks as their input specs.

Comment (emailweixu) [reply alf/networks/network.py]: Perhaps change to a different name so not to be confused? For example, processed_tensor_spec

File: alf/networks/network.py (hunk 14 above)
Comment (hnyu) [reply alf/networks/network.py]: If `input_preprocessors` is None, the user can still use the combiner (e.g., concat)

File: alf/algorithms/diayn_algorithm.py [hunk 19]
```diff
@@ -106,39 +117,40 @@ def train_step(self,
...
+        skill_pred, _ = self._discriminator_net(feature)
 
-        skill_discriminate_loss = tf.nn.softmax_cross_entropy_with_logits(
-            labels=prev_skill, logits=skill_pred)
+        if self._skill_spec.is_discrete:
+            loss = torch.nn.CrossEntropyLoss(reduction='none')(
+                input=skill_pred, target=prev_skill.to(torch.int64))
+        else:
+            # nn.MSELoss doesn't support reducing along a dim
+            loss = torch.sum(math_ops.square(skill_pred - prev_skill), dim=-1)
```
Comment (Haichao-Zhang) [inline alf/algorithms/diayn_algorithm.py]: Might be better to use ```mean``` for reducing. ```sum``` is more easily impacted by the dims of the skill vectors.

File: alf/networks/actor_distribution_networks.py [hunk 20]
```diff
@@ -149,11 +199,11 @@ def forward(self, observation, state):
             act_dist (torch.distributions): action distribution
             new_state (nest[tuple]): the updated states
         """
-        encoding = self._encoding_net(observation)
-        encoding, state = self._lstm_encoding_net(encoding, state)
-        act_dist = self._projection_net(encoding)
+        observation, state = Network.forward(self, observation, state)
```
Comment (Haichao-Zhang) [inline alf/networks/actor_distribution_networks.py]: Is it possible to wrap this call (```Network.forward```) into a function like ```input_processing(...)``` in base ```Network``` class?  Then this line would be like ```observation, state = self.input_processing(self, observation, state)```, which is more natural and clear in terms what it does. 
The user can also override this ```input_processing``` function in their own class if customization is desirable.

File: alf/algorithms/diayn_algorithm.py (hunk 19 above)
Comment (hnyu) [reply alf/algorithms/diayn_algorithm.py]: It's nothing wrong with sum because it's essentially the log likelihood of a normal distribution with a fixed std.

File: alf/networks/actor_distribution_networks.py (hunk 20 above)
Comment (hnyu) [reply alf/networks/actor_distribution_networks.py]: I put it in `Network.forward` because that can remind the user of calling it to some extent. Basically it's always beneficial to first call this before you do other things.
